### PR TITLE
Crm 14920 sort order labels custom search 2nd try

### PR DESCRIPTION
**Pull request by johanv**
This PR replaces #7446. It seems to fix CRM-14920, but it uses a dodgy workaround for CRM-16587.

---
- [CRM-14920: Custom Searches do not honor user sort order](https://issues.civicrm.org/jira/browse/CRM-14920)
- [CRM-16587: Action menu is not available on selected records on a Custom search](https://issues.civicrm.org/jira/browse/CRM-16587)

### CODE REVIEW

**Comment by davecivicrm** [comment]
Test failures are unrelated.

**Comment by davecivicrm** [comment]
@johanv The sort is working now on several searches (Postal Mailing, Proximity) - yay. However, in my testing the "select rows" => "action menu enabled" flow breaks with this patch (e.g. causes a regression against CRM-16587). Screenshot attached:
screen shot 2015-12-15 at 11 25 47 am

**Comment by johanv** [comment]
The problem you describe only occurs when you sort on Contact ID, sorting on any other column would work.
Now if you would try the custom search 'zip code range', you would have the problem on the current master branch as well (so it is not completely caused  by my patch ;)).
After my last commit, you can now always select results for PostMailing and ZipCodeRange, but sorting on Contact-ID fails for people without address for PostMailing.
So there is still work to do.

**Comment by davecivicrm** [comment]
@johanv Tested with the updated PR and the select rows problem is fixed. However, there's another regression:
- Postal Mailing Search
- Sort by State
- Select several rows
- Task = "Group - add contacts" (and them to some group)
When the search results reload, the column names/sort arrows are missing from the top of the selector table.

This does not happen without the PR applied.
screen shot 2015-12-15 at 3 37 41 pm

File: CRM/Contact/Form/Task.php [hunk 1]
```diff
@@ -190,9 +190,7 @@ public static function preProcessCommon(&$form, $useTable = FALSE) {
         // and it decides when to use distinct based on input criteria, which needs
         // to be fixed and optimized.
 
-        foreach ($allCids[$cacheKey] as $cid => $ignore) {
-          $form->_contactIds[] = $cid;
-        }
+        $form->_contactIds = array_unique($allCids[$cacheKey]);
```
Comment (davecivicrm) [inline CRM/Contact/Form/Task.php]: Can you comment on why this change is needed. This function is shared by lots of code paths so I want to understand why this change is needed. Same for change in line 314.

Comment (johanv) [inline CRM/Contact/Form/Task.php]: This was my first attempt to fix the issue, but I guess it is nod needed. I will revert this.

**Comment by davecivicrm** [comment]
@johanv This PR slightly aggravates the existing problem with column headers disappearing - https://issues.civicrm.org/jira/browse/CRM-17724- (at least for Postal Mailing) since the State header ALSO disappears after a task. However, I think it leaves things more usable than w/o it so I'm merging.
